### PR TITLE
Relaxed nnz requirement on serial banded matrices

### PR DESCRIPTION
**Pull request by mattkab2**
The serial banded matrix requires an exact nnz value count to function. However, this was improved for the parallel versions, which instead treat nnz as an upper bound.

This MR updates the serial behavior to match the parallel behavior, which resolves several ongoing problems with implementing native CMFD in the serial debug test cases.

### CODE REVIEW

File: src/MatrixTypes_Native.f90
Comment (djabaay) [inline src/MatrixTypes_Native.f90:694]: Would the requirement be for `thisMatrix%nnz >= thisMatrix%counter` ?

File: src/MatrixTypes_Native.f90
Comment (djabaay) [inline src/MatrixTypes_Native.f90:698]: What's the significance of changing from `thisMatrix%iTmp` to `thisMatrix%counter`?

Comment (mattkab2) [reply src/MatrixTypes_Native.f90:698]: Essentially this means it will loop over the added entries instead of the full iTmp array.
iTmp is an allocated array which stores up to nnz entries, but now because of these changes it may not be full.
So we loop up to counter, which counts the number of entries added prior to the assemble call (which makes the matrix structure read-only)

File: src/MatrixTypes_Native.f90
Comment (mattkab2) [reply src/MatrixTypes_Native.f90:694]: Correct, all though this requirement is better placed in the set routine, because otherwise an array access error will happen before the requirement is called (counter is equal to the index of the last set element in iTmp).

I'll make the change and push up.

File: src/MatrixTypes_Native.f90
Comment (djabaay) [reply src/MatrixTypes_Native.f90:698]: Got it, so `%counter` is the number of entries set on the `%iTmp` array.

File: unit_tests/testMatrixTypes/testMatrixTypes.f90
Comment (djabaay) [inline unit_tests/testMatrixTypes/testMatrixTypes.f90:1444]: Delete the debug statement

File: unit_tests/testMatrixTypes/testMatrixTypes.f90
Comment (djabaay) [inline unit_tests/testMatrixTypes/testMatrixTypes.f90:1419]: Delete the debug statement

File: src/MatrixTypes_Native.f90
Comment (djabaay) [inline src/MatrixTypes_Native.f90:1373]: `<` or `<=` ?

Comment (mattkab2) [reply src/MatrixTypes_Native.f90:1373]: It needs to be < because counter is incremented before the set call. Otherwise counter would equal the number of entries plus one, which isn't a disaster but I chose to set it up this way.

Comment (djabaay) [reply src/MatrixTypes_Native.f90:1373]: Perhaps just make a note of that somewhere if you haven't already.

File: unit_tests/testMatrixTypes/testMatrixTypes.f90
Comment (mattkab2) [reply unit_tests/testMatrixTypes/testMatrixTypes.f90]: Removed

File: unit_tests/testMatrixTypes/testMatrixTypes.f90
Comment (mattkab2) [reply unit_tests/testMatrixTypes/testMatrixTypes.f90]: Removed

**Comment by djabaay** [comment]
Quoting @mattkab2 from our offline conversation in regards to changing the DBC from `<=` to `<`:

I guess it would be more accurate to say nnz is the max nonzeros the matrix can hold.
The counter starts at zero when initialized
11:21
Then when set is called, it is incremented and iTmp(counter) gets set
11:21
But iTmp is allocated to have nnz entries, so counter can never be equal to nnz at the start of the set call or there will be a memory access error
11:22
At least, before the assemble routine is called.
After assemble is called it is assumed that you're modifying an existing entry

This fully explains what is going on, and makes sense to me.

@aarograh Do you want to take a quick look through before merging?

**Review by djabaay — APPROVED** [review]

**Comment by mattkab2** [comment]
@aarograh Mind taking a quick look?

**Comment by aarograh** [comment]
@mattkab2 I don't see any issues that need to be addressed with these changes.  We're tagging VERA 4.3-RC1 soon, so I'm going to hold off merging this until after the tag just to be on the safe side.  I'll merge as soon as I know we've tagged.

**Comment by aarograh** [comment]
Sounds like things are going a bit slower with the tag than I realized, so I'm going ahead with this.  We have time to back it out if something were to go wrong.

**Pull request merged**
